jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java

```diff
@@ -0,0 +1,459 @@
+import org.eclipse.jetty.server.ForwardedRequestCustomizer;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.IncludeExcludeSet;
+import org.eclipse.jetty.util.InetAddressSet;
+import org.eclipse.jetty.util.StringUtil;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+import org.eclipse.jetty.util.annotation.Name;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.thread.Locker;
+import org.eclipse.jetty.util.thread.ThreadPool;
+import org.eclipse.jetty.util.thread.ThreadPool.SizedThreadPool;
+
+
+/**
+ * <p>Handler to limit the threads per IP address for DOS protection</p>
+ * <p>The ThreadLimitHandler applies a limit to the number of Threads 
+ * that can be used simultaneously per remote IP address.
+ * </p>
+ * <p>The handler makes a determination of the remote IP separately to
+ * any that may be made by the {@link ForwardedRequestCustomizer} or similar:
+ * <ul>
+ * <li>This handler will use either only a single style
+ * of forwarded header.   This is on the assumption that a trusted local proxy
+ * will produce only a single forwarded header and that any additional
+ * headers are likely from untrusted client side proxies.</li>
+ * <li>If multiple instances of a forwarded header are provided, this
+ * handler will use the right-most instance, which will have been set from
+ * the trusted local proxy</li>
+ * </ul>
+ * Requests in excess of the limit will be asynchronously suspended until
+ * a thread is available.  
+ * <p>This is a simpler alternative to DosFilter</p>
+ */
+public class ThreadLimitHandler extends HandlerWrapper
+{
+    private static final Logger LOG = Log.getLogger(ThreadLimitHandler.class);
+
+    private final static String REMOTE = "o.e.j.s.h.TLH.REMOTE";
+    private final static String PASS = "o.e.j.s.h.TLH.PASS";
+    private final boolean _rfc7239;
+    private final String _forwardedHeader;
+    private final IncludeExcludeSet<String, InetAddress> _includeExcludeSet = new IncludeExcludeSet<>(InetAddressSet.class);
+    private final ConcurrentMap<String, Remote> _remotes = new ConcurrentHashMap<>();
+    private volatile boolean _enabled;
+    private long _blockForMs=0;
+    private int _threadLimit=10;
+
+    public ThreadLimitHandler()
+    {
+        this(null,false);
+    }
+    
+    public ThreadLimitHandler(@Name("forwardedHeader") String forwardedHeader)
+    {
+        this(forwardedHeader,HttpHeader.FORWARDED.is(forwardedHeader));
+    }
+
+    public ThreadLimitHandler(@Name("forwardedHeader") String forwardedHeader, @Name("rfc7239")  boolean rfc7239)
+    {
+        super();
+        _rfc7239 = rfc7239;
+        _forwardedHeader = forwardedHeader;
+        _enabled = true;
+    }
+    
+    @Override
+    protected void doStart() throws Exception
+    {
+        if (_threadLimit<0)
+            _enabled=false;
```

Comment:
Let's have just one mechanism to enable/disable this handler.
Otherwise when done via JMX we must remember to first set the `threadLimit` to a positive number and _then_ enable it. 
I suggest we only have `threadLimit` and if it's negative it's disabled.